File: combine_translations.js
const fs = require('fs');
const path = require('path');

const LANGS_DIR = 'langs';
const OUT_DIR = 'data/translations/chapters';

const ALL_LANGUAGES = ['am', 'apc', 'apd', 'ar', 'arz', 'bho', 'bn', 'ceb', 'cs', 'da', 'de', 'es', 'fa', 'fi', 'fr', 'gu', 'ha', 'hi', 'id', 'it', 'ja', 'jv', 'km', 'kn', 'ko', 'mai', 'ml', 'mr', 'ms', 'my', 'nl', 'no', 'or', 'pa', 'pl', 'pt', 'ru', 'sv', 'sw', 'ta', 'te', 'th', 'tl', 'tr', 'uk', 'ur', 'uz', 'vi', 'wuu', 'yo', 'yue', 'zh'];

const args = process.argv.slice(2);
let LANGUAGES = [];

if (args.length > 0) {
    LANGUAGES = args;
    console.log(`Building only: ${LANGUAGES.join(', ')}`);
} else {
    LANGUAGES = ALL_LANGUAGES;
    console.log(`Building ALL languages (${LANGUAGES.length})`);
}

function combineLang(iso) {
    console.log(`Processing ${iso}...`);
    const dirPath = path.join(LANGS_DIR, iso);
    if (!fs.existsSync(dirPath)) {
        console.error(`Directory not found: ${dirPath}`);
        return;
    }

    const files = fs.readdirSync(dirPath)
        .filter(f => f.endsWith('.json'))
        .sort();

    if (files.length === 0) {
        console.error(`No JSON files found for ${iso}`);
        return;
    }

    let allChapters = [];
    for (const f of files) {
    try {
        const content = fs.readFileSync(path.join(dirPath, f), 'utf8').replace(/^\uFEFF/, '');
        const data = JSON.parse(content);
        if (Array.isArray(data)) {
            allChapters = allChapters.concat(data);
        } else {
            allChapters.push(data);
        }
    } catch (e) {
        console.error(`Error reading ${f}: ${e}`);
    }
    }

    allChapters.sort((a, b) => (a.num || 0) - (b.num || 0));
    allChapters = allChapters.filter((c, i, arr) => !i || c.num !== arr[i-1].num);
    console.log(`Total chapters for ${iso}: ${allChapters.length}`);

    if (!fs.existsSync(OUT_DIR)) {
        fs.mkdirSync(OUT_DIR, { recursive: true });
    }

    const outFile = path.join(OUT_DIR, `${iso}.json`);
    fs.writeFileSync(outFile, JSON.stringify(allChapters, null, 2), 'utf8');
    console.log(`Written to ${outFile}`);
}

LANGUAGES.forEach(combineLang);

console.log('\nDone!');

File: combine_translations.py
import json
import os
import glob

LANGUAGES = ['de', 'it', 'ja', 'ko', 'pt', 'tr']
LANGS_DIR = 'langs'
OUT_DIR = 'data/translations'

def combine_lang(iso):
    print(f"Processing {iso}...")
    path = os.path.join(LANGS_DIR, iso, "*.json")
    files = sorted(glob.glob(path))
    
    if not files:
        print(f"No files found for {iso}")
        return

    all_chapters = []
    for f in files:
        try:
            with open(f, 'r', encoding='utf-8') as fp:
                data = json.load(fp)
                if isinstance(data, list):
                    all_chapters.extend(data)
                else:
                    all_chapters.append(data)
        except Exception as e:
            print(f"Error reading {f}: {e}")

    # Sort by chapter number
    all_chapters.sort(key=lambda x: x.get('num', 0))
    
    print(f"Total chapters for {iso}: {len(all_chapters)}")
    
    out_file = os.path.join(OUT_DIR, f"chapters_{iso}.json")
    with open(out_file, 'w', encoding='utf-8') as fp:
        json.dump(all_chapters, fp, ensure_ascii=False, indent=2)
    print(f"Written to {out_file}")

if __name__ == "__main__":
    os.makedirs(OUT_DIR, exist_ok=True)
    for iso in LANGUAGES:
        combine_lang(iso)
